lib/upnp/igd_service.go

```diff
@@ -49,12 +50,113 @@ type IGDService struct {
+	var returnErr error = nil
+	var successfulIPs []net.IP
+	if s.Interface == nil {
+		return nil, errors.New("no interface")
+	}
+
+	addrs, err := s.Interface.Addrs()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+
+		if err != nil {
+			l.Infoln("Couldn't parse address", addr, err)
+			continue
+		}
+
+		// IsGlobalUnicast returns true for ULAs so these must be checked for seperately
+		if ip.To4() != nil || !ip.IsGlobalUnicast() || ip.IsPrivate() {
+			continue
+		}
+
+		err = s.tryAddPinholeForIP6(ctx, protocol, port, duration, ip)
+		if err != nil {
+			l.Infof("Couldn't add pinhole for [%s]:%d/%s. %s", ip.String(), port, protocol, err.Error())
+			returnErr = err
+		} else {
+			successfulIPs = append(successfulIPs, ip)
+		}
+	}
+
+	if len(successfulIPs) > 0 {
+		// (Maybe partial) success, we added a pinhole for at least one GUA.
+		return successfulIPs, nil
+	} else {
+		return nil, returnErr
+	}
+}
+
+func (s *IGDService) tryAddPinholeForIP6(ctx context.Context, protocol nat.Protocol, port int, duration time.Duration, ip net.IP) error {
+	var protoNumber int
+	if protocol == nat.TCP {
+		protoNumber = 6
+	} else if protocol == nat.UDP {
+		protoNumber = 17
+	} else {
+		return errors.New("protocol not supported")
+	}
+
+	const template = `<u:AddPinhole xmlns:u="%s">
+	<RemoteHost></RemoteHost>
+	<RemotePort>0</RemotePort>
+	<Protocol>%d</Protocol>
+	<InternalPort>%d</InternalPort>
+	<InternalClient>%s</InternalClient>
+	<LeaseTime>%d</LeaseTime>
+	</u:AddPinhole>`
+
+	body := fmt.Sprintf(template, s.URN, protoNumber, port, ip.String(), duration/time.Second)
+
+	// IP should be a global unicast address, so we can use it as the source IP.
+	// By the UPnP spec, the source address for unauthenticated clients should be the same as the InternalAddress the pinhole is requested for.
+	// Currently, WANIPv6FirewallProtocol is restricted to IPv6 gateways, so we can always set the IP.
+	var err error
+	var resp []byte
+	resp, err = soapRequestWithIP(ctx, s.URL, s.URN, "AddPinhole", body, &net.TCPAddr{IP: ip})
+	succResponse := &soapAddPinholeResponse{}
+	envelope := &soapErrorResponse{}
+
+	if err != nil && resp != nil {
+		if unmarshalErr := xml.Unmarshal(resp, envelope); unmarshalErr != nil {
+			// There is an error response that we cannot parse.
+			return unmarshalErr
+		} else {
+			// There is a parsable UPnP error. Return that.
+			return fmt.Errorf("UPnP error: %s (%d)", envelope.ErrorDescription, envelope.ErrorCode)
+		}
+	} else if resp != nil {
+		// Ignore errors since this is only used for debug logging.
+		unmarshalErr := xml.Unmarshal(resp, succResponse)
+		if unmarshalErr == nil {
+			l.Debugf("UPnPv6: UID for pinhole on [%s]:%d/%s is %d", ip.String(), port, protocol, succResponse.UniqueID)
+		} else {
+			l.Debugln("Failed to parse respone from gateway")
+		}
```

Comment:
* Some simplifications on the conditional
* Use the more tightly scoped success response var
* Follow the usual code pattern of checking the error case first
* Include the unmarshal error in the debug message
* Again `.String()` vs. `%s`
* Include gateway `ID()` in debug messages
```suggestion
		var succResponse soapAddPinholeResponse
		if unmarshalErr := xml.Unmarshal(resp, &succResponse); unmarshalErr != nil {
			// Ignore errors since this is only used for debug logging.
			l.Debugf("Failed to parse response from gateway %s: %s", s.ID(), unmarshalErr)
		} else {
			l.Debugf("UPnPv6: UID for pinhole on [%s]:%d/%s is %d on gateway", ip, port, protocol, succResponse.UniqueID, s.ID())
		}
```